lib/hanami/view/html.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+module Hanami
+  class View
+    module HTML
+      # A string that has been marked as "HTML safe", ensuring that it is not automatically escaped
+      # when used in HTML view templates.
+      #
+      # A SafeString is frozen when initialized to ensure it cannot be mutated after being marked as
+      # safe, which could be an avenue for injection of unsafe content.
+      #
+      # @see String#html_safe
+      #
+      # @api public
+      # @since 2.0.0
+      class SafeString < String
+        # @api public
+        # @since 2.0.0
+        def initialize(string)
+          super(string)
+          freeze
+        end
+
+        # @return [true]
+        #
+        # @api public
+        # @since 2.0.0
+        def html_safe?
+          true
+        end
+
+        # @return [self]
+        #
+        # @api public
+        # @since 2.0.0
+        def html_safe
+          self
+        end
+
+        # @return [self]
+        #
+        # @api public
+        # @since 2.0.0
+        def to_s
+          self
+        end
```

Comment:
```suggestion
        end


        private

        def initialize_copy(other)
          super
          freeze
        end
```

I forgot to point you need to freeze copies as well, otherwise:

```ruby
str = "foo".html_safe.dup
p str.html_safe? # => true
str << "<script>"
p str # => "foo<script>"

```

Reply:
Arf, even that won't be enough, many string methods will return you a `SafeString`:

```ruby
str = "foo".html_safe
p str.sub("foo", "<script>").class # Hanami::View::HTML::SafeString
```

So delegation is likely the way to go here. e.g. `class SafeString < DelegateClass(String)`

Reply:
@byroot Thanks for circling back here! Great tip about the initialize_copy. I've just added this in https://github.com/hanami/view/pull/226/commits/4647c40125ffd9dcbcaf95bf21a992e15dc54966.

As for your `str.sub` example, I actually tested that during my earlier work here, and even included a test case:

```ruby
  it "returns non-safe strings from methods returning new strings" do
    new_string = safe_string.gsub(/hello/, "world")
    expect(new_string).to eq "world"
    expect(new_string).not_to be_html_safe
  end
```

The test passed so I presumed things to be fine...

If I check things out more manually, here's what I'm seeing:

```ruby
safe_str = "Hello".html_safe
# => "Hello"
safe_str.html_safe?
# => true
safe_str.class
# => Hanami::View::HTML::SafeString

new_str = safe_str.gsub("l", "y")
# => "Heyyo"
new_str.html_safe?
# => false
new_str.class
# => String
```

This is consistent with what the test was exercising... but maybe I'm missing something? Is there another category of string methods that when called on a subclass return a new instance of that subclass rather than a plain `String` itself?

Reply:
> Is there another category of string methods that when called on a subclass return a new instance of that subclass rather than a plain String itself?

Nah, my bad, I was testing on 2.7, the behavior of Ruby this changed in 3.0. Hamani being 3.0+ you're good.

Sorry for the false alarm.

Reply:
> I was testing on 2.7, the behavior of Ruby this changed in 3.0.

TIL! And no worries at all. I appreciate all your input here! ❤️

Reply:
For completeness:

  - https://bugs.ruby-lang.org/issues/10845
  - https://github.com/ruby/ruby/pull/3701

It was indeed changed for all String methods except two, but that are safe in this context (`@+` and `@-`).